pyro/nn/module.py

```diff
@@ -121,8 +121,8 @@ def _pyro_set_supermodule(self, name, cache):
         self._pyro_name = name
         self._pyro_cache = cache
         for key, value in self._modules.items():
-            assert isinstance(value, PyroModule)
-            value._pyro_set_supermodule(_make_name(name, key), cache)
+            if isinstance(value, PyroModule):
+                value._pyro_set_supermodule(_make_name(name, key), cache)
```

Comment:
@fritzo I think all the tests have passed with only minor changes in the codes. The API is almost the same as previously (you can take a look at GP tutorial, where we only need to change `set_prior` to `PyroSample`). I think that most of the changes would be at your expectations. There are two points which might surprise you so I would like to point out before working with docs:
+ Method `set_mode` is a context manager now (to cache values) without having to run `__call__`.
+ I have relaxed the assertation here. Actually, I don't quite understand the side effects of this change, but without this, [sv-dkl example](https://github.com/pyro-ppl/pyro/blob/dev/examples/contrib/gp/sv-dkl.py) raises an error here. I guess the reason is [CNN](https://github.com/pyro-ppl/pyro/blob/dev/examples/contrib/gp/sv-dkl.py#L42) is a nn.Module instead of `PyroModule`?

Reply:
Thanks for fixing this, I am glad we are working out issues of `PyroModule` before the next release!